### ISSUE TITLE
CoreCLR crashes when run with an unlimited stack size on Linux

### ISSUE THREAD

**Issue by jaredpar** [dotnet/coreclr#2467]
Recently Roslyn transitioned it's developer build to be makefile based.  The default `make` install on Ubuntu has a bug / feature where by child processes are created with an unlimited stack size (default on Ubuntu is 8192).  

An unlimited stack size seems to cause issues for CoreCLR.  Virtually all invocations of Roslyn in this context crash with an exception.  Running the exact same Roslyn invocation outside of make, or from `make` after explicitly resetting the stack size to 8192, succeed.  

Generally the exception is an access violation allocating an array [at this location](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/SyntaxListBuilder.cs#L17).  But it also shows up as a `PALException` in other locations.  

Repro steps:

- Clone [roslyn](https://github.com/dotnet/roslyn) on Linux (won't repro on Mac).
- Comment out the ulimit line in [this file](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/CscCore/csc#L8) 
- Run `make bootstrap`
- Run `make BOOTSTRAP=true`

The majority of csc invocations from the last `make` command will fail with the following error:

``` txt
error MSB6006: "csc" exited with code 134. [/home/jaredpar/code/roslyn/src/Debugging/Microsoft.DiaSymReader.PortablePdb/Microsoft.DiaSymReader.PortablePdb.csproj]
```

Please let me know if there are any other details I can provide here to help track this down.

**Comment by janvorli** [comment]
@jaredpar I'll take a look. Sounds like a problem getting stack limits in the CoreCLR PAL in case the ulimit specifies stack as unlimited.

**Comment by janvorli** [comment]
@jaredpar I can see that the stack limits are obtained correctly in the PAL. It is interesting though that when the ulimit for stack is set to unlimited, the pthreads set the stack to 2MB. When the ulimit is set to a specific value, pthreads set the stack to that value. The default ulimit value 8192 (which means 8MB) is larger than the "unlimited" in this case. So the issue is likely caused by too small stack in your case.

**Comment by janvorli** [comment]
I keep investigating it.

**Comment by janvorli** [comment]
@jaredpar I have tried that locally and the csc script doesn't seem to be invoked. I get the same error no matter whether I modify it or not. I've also tried to add gdb invocation into the script and nothing has changed.

**Comment by jaredpar** [comment]
@janvorli okay i think i understand the confusion.  The `csc` file will be moved to `Binaries/Bootstrap/csc` during the `make bootstrap` action.  That is the location that will get invoked when running `make BOOTSTRAP=true` action.  Sorry for the confusion.

**Comment by janvorli** [comment]
After more investigation, I have found that while pthread sets the stack limit of all secondary threads to 2MB when unlimited is specified, the main thread's limit is initially really set to "unlimited", which on my machine is about 85TB. However, later on (maybe after the first secondary thread is launched), the limit of the main thread changes a bit (by 512MB). But it is too late, CoreCLR has already cached the "wrong" value. 
Later on, the C++ runtime places various data structures and shared libraries into the memory range that was originally reserved for the stack, including stacks of the secondary threads.
The Thread::IsAddressInStack then returns incorrect results which leads into various strange failures.

**Comment by janvorli** [comment]
Unfortunatelly, it looks like there is no way to have the stack of the main thread bound by a static limit in case of `ulimit -s` set to unlimited. Although pthreads have a function pthread_attr_setstacksize, it only sets the minimum size of stack and the main thread is allowed to grow its stack dynamically. Only the secondary threads have fixed sized stacks.
The only solution seems to be to set the limit to some arbitrary value by calling setrlimit in the PAL initialization in case the limit is set to unlimited. 
Not caching the stack limit and reading it dynamically is not possible both due to perf reasons and the fact that during GC stack scan, we need to know the current value for other than the current thread.
@jkotas any thoughts on that?

**Comment by jkotas** [comment]
Is there a way to tell at runtime that the thread has unlimited stack size?

> during GC stack scan, we need to know the current value for other than the current thread

We do not actually even need the actual stack limit in this case, we really just need the current managed SP to get the current managed stack range ... that is available to the GC stack scan.

Most of the other places where we check stack limits seems to be debug-only checks -  it may be ok to make them slower.

**Comment by janvorli** [comment]
Fixed by #2573
